src/client/serialize/hdf_serialize.c

```diff
@@ -0,0 +1,1034 @@
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_LAYOUT_VER");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[3];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_CSUM");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[4];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_CSUM_CHUNK_SIZE");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[5];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_CSUM_SERVER_VERIFY");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[6];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_REDUN_FAC");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[7];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_REDUN_LVL");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[8];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_SNAPSHOT_MAX");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[9];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_COMPRESS");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[10];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_ENCRYPT");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[11];
+	rc = cont_serialize_prop_str(file_id, entry, "DAOS_PROP_CO_OWNER");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[12];
+	rc = cont_serialize_prop_str(file_id, entry,
+				     "DAOS_PROP_CO_OWNER_GROUP");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[13];
+	rc = cont_serialize_prop_uint(file_id, entry, "DAOS_PROP_CO_DEDUP");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[14];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_DEDUP_THRESHOLD");
+	if (rc != 0) {
+		goto out;
+	}
+
+	entry = &prop_query->dpp_entries[15];
+	rc = cont_serialize_prop_uint(file_id, entry,
+				      "DAOS_PROP_CO_ALLOCED_OID");
+	if (rc != 0) {
+		goto out;
+	}
+
+	/* serialize ACL */
+	if (prop_query->dpp_nr > 16) {
+		entry = &prop_query->dpp_entries[16];
+		rc = cont_serialize_prop_acl(file_id, entry,
+					     "DAOS_PROP_CO_ACL");
+		if (rc != 0) {
+			goto out;
+		}
+	}
+
+out:
+	return rc;
+}
+
+int
+cont_serialize_usr_attrs(hid_t *file_id, hid_t *usr_attr_memtype, int num_attrs,
+			 char **names, char **buffers, size_t *sizes,
+			 daos_handle_t coh)
+{
+	int		rc = 0;
+	hid_t		status = 0;
+	hid_t		dset = 0;
+	hid_t		dspace = 0;
+	hsize_t		dims[1];
+	struct usr_attr	*attr_data = NULL;
+	int		i;
+
+	if (num_attrs == 0) {
+		goto out_no_attrs;
+	}
+
+	/* Create the user attribute data space */
+	dims[0] = num_attrs;
+	dspace = H5Screate_simple(1, dims, NULL);
+	if (dspace < 0) {
+		fprintf(stderr, "failed to create user attr dspace");
+		rc = 1;
+		goto out;
+	}
```

Comment:
In general, for this API, I'm not sure what we want the return codes to be. In mpifileutils we use generic 1/-1 for a lot of things, but in DAOS the standard is to use the -DER codes. We might need to propagate errors up from helper functions and return the corresponding -DER codes, or MISC/generic for HDF5 errors.
Probably best to double check with someone who knows better.